File: src/ee/app_lovin/private/AppLovinRewardedAdListener.swift
//
//  AppLovinRewardedAdListener.swift
//  Pods
//
//  Created by eps on 6/26/20.
//

import AppLovinSDK

private let kTag = "\(AppLovinRewardedAdListener.self)"
private let kPrefix = "AppLovinBridge"
private let kOnRewardedAdLoaded = "\(kPrefix)OnRewardedAdLoaded"
private let kOnRewardedAdFailedToLoad = "\(kPrefix)OnRewardedAdFailedToLoad"
private let kOnRewardedAdClicked = "\(kPrefix)OnRewardedAdClicked"
private let kOnRewardedAdClosed = "\(kPrefix)OnRewardedAdClosed"

internal class AppLovinRewardedAdListener:
    NSObject, ALAdLoadDelegate, ALAdDisplayDelegate, ALAdRewardDelegate, ALAdVideoPlaybackDelegate {
    private let _bridge: IMessageBridge
    private let _logger: ILogger
    private var _isLoaded = false
    private var _rewarded = false

    init(_ bridge: IMessageBridge, _ logger: ILogger) {
        _bridge = bridge
        _logger = logger
        super.init()
    }

    var isLoaded: Bool {
        return _isLoaded
    }

    func adService(_ adService: ALAdService, didLoad ad: ALAd) {
        _logger.debug("\(kTag): \(#function)")
        _isLoaded = true
        _bridge.callCpp(kOnRewardedAdLoaded)
    }

    func adService(_ adService: ALAdService, didFailToLoadAdWithError code: Int32) {
        _logger.debug("\(kTag): \(#function): code \(code)")
        _bridge.callCpp(kOnRewardedAdFailedToLoad, "\(code)")
    }

    func ad(_ ad: ALAd, wasDisplayedIn view: UIView) {
        _logger.debug("\(kTag): \(#function)")
        _isLoaded = false
    }

    func videoPlaybackBegan(in ad: ALAd) {
        _logger.debug("\(kTag): \(#function)")
    }

    func ad(_ ad: ALAd, wasClickedIn view: UIView) {
        _logger.debug("\(kTag): \(#function)")
        _bridge.callCpp(kOnRewardedAdClicked)
    }

    func videoPlaybackEnded(in ad: ALAd, atPlaybackPercent percentPlayed: NSNumber, fullyWatched wasFullyWatched: Bool) {
        _logger.debug("\(kTag): \(#function)")
    }

    func rewardValidationRequest(for ad: ALAd, didSucceedWithResponse response: [AnyHashable: Any]) {
        _logger.debug("\(kTag): \(#function)")
        _rewarded = true
    }

    func rewardValidationRequest(for ad: ALAd, wasRejectedWithResponse response: [AnyHashable: Any]) {
        _logger.debug("\(kTag): \(#function)")
    }

    func rewardValidationRequest(for ad: ALAd, didExceedQuotaWithResponse response: [AnyHashable: Any]) {
        _logger.debug("\(kTag): \(#function)")
    }

    func rewardValidationRequest(for ad: ALAd, didFailWithError responseCode: Int) {
        _logger.debug("\(kTag): \(#function): code \(responseCode)")
    }

    func ad(_ ad: ALAd, wasHiddenIn view: UIView) {
        _logger.debug("\(kTag): \(#function)")
        _bridge.callCpp(kOnRewardedAdClosed, Utils.toString(_rewarded))
    }
}

File: src/android/app_lovin/src/main/java/com/ee/internal/AppLovinRewardedAdListener.kt
package com.ee.internal

import com.applovin.sdk.AppLovinAd
import com.applovin.sdk.AppLovinAdClickListener
import com.applovin.sdk.AppLovinAdDisplayListener
import com.applovin.sdk.AppLovinAdLoadListener
import com.applovin.sdk.AppLovinAdRewardListener
import com.applovin.sdk.AppLovinAdVideoPlaybackListener
import com.ee.ILogger
import com.ee.IMessageBridge
import com.ee.Thread
import com.ee.Utils
import kotlinx.serialization.Serializable
import java.util.concurrent.atomic.AtomicBoolean

internal class AppLovinRewardedAdListener(
    private val _bridge: IMessageBridge,
    private val _logger: ILogger)
    : AppLovinAdLoadListener, AppLovinAdDisplayListener, AppLovinAdClickListener, AppLovinAdRewardListener, AppLovinAdVideoPlaybackListener {
    @Serializable
    @Suppress("unused")
    private class ErrorResponse(
        val code: Int,
        val message: String
    )

    companion object {
        private val kTag = AppLovinRewardedAdListener::class.java.name
        private const val kPrefix = "AppLovinBridge"
        private const val kOnRewardedAdLoaded = "${kPrefix}OnRewardedAdLoaded"
        private const val kOnRewardedAdFailedToLoad = "${kPrefix}OnRewardedAdFailedToLoad"
        private const val kOnRewardedAdClicked = "${kPrefix}OnRewardedAdClicked"
        private const val kOnRewardedAdClosed = "${kPrefix}OnRewardedAdClosed"
    }

    private val _isLoaded = AtomicBoolean(false)
    private var _rewarded = false

    val isLoaded: Boolean
        get() = _isLoaded.get()

    override fun adReceived(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::adReceived.name}")
            _isLoaded.set(true)
            _bridge.callCpp(kOnRewardedAdLoaded)
        }
    }

    override fun failedToReceiveAd(errorCode: Int) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::failedToReceiveAd.name}: code $errorCode")
            _bridge.callCpp(kOnRewardedAdFailedToLoad, ErrorResponse(errorCode, "").serialize())
        }
    }

    override fun adDisplayed(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::adDisplayed.name}")
            _rewarded = false
        }
    }

    override fun videoPlaybackBegan(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::videoPlaybackBegan.name}")
        }
    }

    override fun adClicked(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::adClicked.name}")
            _bridge.callCpp(kOnRewardedAdClicked)
        }
    }

    override fun videoPlaybackEnded(ad: AppLovinAd, percentViewed: Double, fullyWatched: Boolean) {
        Thread.runOnMainThread {
            _logger.debug("$kTag: ${this::videoPlaybackEnded.name}")
        }
    }

    override fun userRewardVerified(ad: AppLovinAd, response: Map<String, String>) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::userRewardVerified.name}: $response")
            _rewarded = true
        }
    }

    override fun userRewardRejected(ad: AppLovinAd, response: Map<String, String>) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::userRewardRejected.name}: $response")
        }
    }

    override fun userOverQuota(ad: AppLovinAd, response: Map<String, String>) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::userOverQuota.name}: $response")
        }
    }

    override fun validationRequestFailed(ad: AppLovinAd, responseCode: Int) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::validationRequestFailed.name}: code = $responseCode")
        }
    }

    override fun adHidden(ad: AppLovinAd) {
        Thread.runOnMainThread {
            _logger.info("$kTag: ${this::adHidden.name}")
            _isLoaded.set(false)
            _bridge.callCpp(kOnRewardedAdClosed, Utils.toString(_rewarded))
        }
    }
}